galley/pkg/config/analysis/analyzers/auth/mtls.go

```diff
@@ -0,0 +1,333 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"istio.io/api/authentication/v1alpha1"
+	"istio.io/istio/galley/pkg/config/analysis/analyzers/util"
+
+	v1 "k8s.io/api/core/v1"
+	k8s_labels "k8s.io/apimachinery/pkg/labels"
+
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/analyzers/auth/mtls"
+	"istio.io/istio/galley/pkg/config/meta/metadata"
+	"istio.io/istio/galley/pkg/config/meta/schema/collection"
+	"istio.io/istio/galley/pkg/config/resource"
+)
+
+// MTLSAnalyzer checks for misconfigurations of MTLS policy when autoMtls is
+// disabled. More specifically, it detects situations where a DestinationRule's
+// MTLS usage is in conflict with mTLS specified by a policy.
+//
+// The most common situations that this detects are: 1. A MeshPolicy exists that
+// requires mTLS, but no global destination rule
+//    says to use mTLS.
+// 2. mTLS is used throughout the mesh, but a DestinationRule is added that
+//    doesn't specify mTLS (usually because it was forgotten).
+//
+// The analyzer tries to act more generally by imagining service-to-service
+// traffic and detecting whether or not there's a conflict with regards to mTLS
+// policy. This means it will also detect explicit misconfigurations as well.
+//
+// Note this is very similar to `istioctl authn tls-check`; however this
+// inspection is all done via analyzing configuration rather than requiring a
+// connection to pilot.
+type MTLSAnalyzer struct{}
+
+// Compile-time check that this Analyzer correctly implements the interface
+var _ analysis.Analyzer = &MTLSAnalyzer{}
+
+// Metadata implements Analyzer
+func (s *MTLSAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name: "auth.MTLSAnalyzer",
+
+		// Each analyzer should register the collections that it needs to use as input.
+		Inputs: collection.Names{
+			metadata.K8SCoreV1Pods,
+			metadata.K8SCoreV1Namespaces,
+			metadata.K8SCoreV1Services,
+			metadata.IstioAuthenticationV1Alpha1Meshpolicies,
+			metadata.IstioAuthenticationV1Alpha1Policies,
+			metadata.IstioNetworkingV1Alpha3Destinationrules,
+		},
+	}
+}
+
+// Analyze implements Analyzer
+func (s *MTLSAnalyzer) Analyze(c analysis.Context) {
+	// TODO Reuse pilot logic as a library rather than reproducing its logic
+	// here.
+
+	// If autoMTLS is turned on, bail out early as the logic used below does not
+	// reason about its usage.
+	autoMtlsEnabled := false
+	c.ForEach(metadata.IstioMeshV1Alpha1MeshConfig, func(r *resource.Entry) bool {
+		mc := r.Item.(*meshconfig.MeshConfig)
+		if mc.GetEnableAutoMtls() != nil && mc.GetEnableAutoMtls().Value {
+			autoMtlsEnabled = true
+		}
+		return true
+	})
```

Comment:
I thought about doing that, but then I'll have to encode logic on where to find it - I felt it was easier to leave the analyzer dumb here.

Is there a performance cost to leaving it this way?

Reply:
Probably not, but it's a bit confusing to read because it implies there can be more than one. For now I think a comment would suffice to clarify the intention.

Reply:
Added a comment.